src/sfdc48/salesforce_base_endpoint.bal

```diff
@@ -0,0 +1,120 @@
+//
+// Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+//
+// WSO2 Inc. licenses this file to you under the Apache License,
+// Version 2.0 (the "License"); you may not use this file except
+// in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+//
+
+import ballerina/http;
+import ballerina/oauth2;
+
+# Salesforce Client object.
+# + salesforceClient - OAuth2 client endpoint
+# + salesforceConfiguration - Salesforce Connector configuration
+public type BaseClient client object {
+    http:Client salesforceClient;
+    SalesforceConfiguration salesforceConfiguration;
+
+    # Salesforce Connector endpoint initialization function.
+    # + salesforceConfig - Salesforce Connector configuration
+    public function __init(SalesforceConfiguration salesforceConfig) {
+        self.salesforceConfiguration = salesforceConfig;
+        // Create OAuth2 provider.
+        oauth2:OutboundOAuth2Provider oauth2Provider = new(salesforceConfig.clientConfig);
+        // Create bearer auth handler using created provider.
+        http:BearerAuthHandler bearerHandler = new(oauth2Provider);
+
+        http:ClientSecureSocket? socketConfig = salesforceConfig?.secureSocketConfig;
+        
+        // Create salesforce http client.
+        if (socketConfig is http:ClientSecureSocket) {
+            self.salesforceClient = new(salesforceConfig.baseUrl, {
+                secureSocket: socketConfig,
+                auth: {
+                    authHandler: bearerHandler
+                }
+            });
+        } else {
+            self.salesforceClient = new(salesforceConfig.baseUrl, {
+                auth: {
+                    authHandler: bearerHandler
+                }
+            });
+        }
+    }
+
+    # Get salesforce bulk API client.
+    # + return - salesforce bulk client
+    public remote function getBulkClient() returns BulkClient {
+        BulkClient bulkClient = new(self.salesforceConfiguration);
```

Comment:
Can't we reuse the same HTTP client inside other clients? without reinitializing it again

Reply:
We can for 2 of the clients. But for the bulk client, we need a new http client since there'll be a different auth handler for that. So instead of exposing the same http client via a getter method, I'm just creating separate clients for simplicity.